tsl/src/chunk.c

```diff
@@ -186,3 +206,889 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type, boo
+ * setting a session variable, thus allowing reads during merges. However,
+ * that can easily lead to deadlocks as shown in isolation tests. Therefore,
+ * use the stricter locking settings by default.
+ */
+static MergeLockUpgrade
+merge_chunks_lock_upgrade_mode(void)
+{
+	const char *lockupgrade =
+		GetConfigOption("timescaledb.merge_chunks_lock_upgrade_mode", true, false);
+
+	if (lockupgrade == NULL)
+		return MERGE_LOCK_ACCESS_EXCLUSIVE;
+
+	if (strcmp("upgrade", lockupgrade) == 0)
+		return MERGE_LOCK_UPGRADE;
+
+	if (strcmp("conditional", lockupgrade) == 0)
+		return MERGE_LOCK_CONDITIONAL_UPGRADE;
+
+	return MERGE_LOCK_ACCESS_EXCLUSIVE;
+}
+
+#if (PG_VERSION_NUM >= 170000 && PG_VERSION_NUM <= 170002)
+/*
+ * Workaround for changed behavior in the relation rewrite code that appeared
+ * in PostgreSQL 17.0, but was fixed in 17.3.
+ *
+ * Merge chunks uses the relation rewrite functionality from CLUSTER and
+ * VACUUM FULL. This works for merge because, when writing into a non-empty
+ * relation, new pages are appended while the existing pages remain the
+ * same. In PG17.0, however, that changed so that existing pages in the
+ * relation were zeroed out. The changed behavior was introduced as part of
+ * this commit:
+ *
+ * https://github.com/postgres/postgres/commit/8af256524893987a3e534c6578dd60edfb782a77
+ *
+ * Fortunately, this was fixed in a follow up commit:
+ *
+ * https://github.com/postgres/postgres/commit/9695835538c2c8e9cd0048028b8c85e1bbf5c79c
+ *
+ * The fix is part of PG 17.3. Howevever, this still leaves PG 17.0 - 17.2
+ * with different behavior.
+ *
+ * To make the merge chunks code work for the "broken" versions we make PG
+ * believe the first rewrite operation is the size of the fully merged
+ * relation so that we reserve the full space needed and then "append"
+ * backwards into the zeroed space (see illustration below). By doing this, we
+ * ensure that no valid data is zeroed out. The downside of this approach is
+ * that there will be a lot of unnecessary writing of zero pages. Below is an
+ * example of what the rewrite would look like for merging three relations
+ * with one page each. When writing the first relation, PG believes the merged
+ * relation already contains two pages when starting the rewrite. These two
+ * existing pages will be zeroed. When writing the next relation we tell PG
+ * that there is only one existing page in the merged relation, and so forth.
+ *
+ *  _____________
+ *  |_0_|_0_|_x_|
+ *  _________
+ *  |_0_|_x_|
+ *  _____
+ *  |_x_|
+ *
+ *  Result:
+ *  _____________
+ *  |_x_|_x_|_x_|
+ *
+ */
+static BlockNumber merge_rel_nblocks = 0;
+static BlockNumber *blockoff = NULL;
+static const TableAmRoutine *old_routine = NULL;
+static TableAmRoutine routine = {};
+
+/*
+ * TAM relation size function to make PG believe that the merged relation
+ * contains as specific amount of existing data.
+ */
+static uint64
+merge_relation_size(Relation rel, ForkNumber forkNumber)
```

Comment:
```suggestion
pg17_workaround_merge_relation_size(Relation rel, ForkNumber forkNumber)
```
I know naming is hard, but we can prefix all functions to make the code clear that everything is related to the workaround.

Reply:
Changed.